### DISCUSSION TITLE
Right-to-Left Languages

### DISCUSSION THREAD

**Discussion by shauleiz** [EdgeTX/edgetx#2111] [Q&A] [answered]
I was asked to help with extending EdgeTX to support **R2L languages** (Arabic, Hebrew, Farsi etc.).
Besides the usual task of adding Fonts and creating a dictionary header file, this task would include deeper changes such as Alignment of the text to the right and Reordering punctuation marks.
I have 30 years of experience with C/C++ but I still fail to find my way in the code. **Is EdgeTX an OS**, and if so, how does in communicate with the HW layer (BIOS?). If it is a higher level then does the underlying OS support R2L display? How?

**Comment by gagarinlg** [comment, marked as answer]
There is an underlying OS, FreeRTOS, but this does not handle any UI.
For UI we are using LVGL from the next release on, which has support for R2L languages, but nobody took any care, so far, to handle this correctly. LVGL is then encapsuled in libopenui.
All texts are using UTF now. When the fonts are generated, a specific set of characters is added, to save space.
You can join the discord server and start asking questions in the #dev channel, for a more real time communication.
